rust/kernel/spi.rs

```diff
@@ -0,0 +1,177 @@
+// SPDX-License-Identifier: GPL-2.0
+
+use crate::bindings;
+use crate::c_types;
+use crate::error::{Error, Result};
+use crate::CStr;
+use alloc::boxed::Box;
+use core::pin::Pin;
+
+#[derive(Clone, Copy)]
+pub struct SpiDevice(*mut bindings::spi_device);
+
+impl SpiDevice {
+    pub unsafe fn from_ptr(dev: *mut bindings::spi_device) -> Self {
+        SpiDevice(dev)
+    }
+
+    pub fn to_ptr(&mut self) -> *mut bindings::spi_device {
+        self.0
+    }
+}
+
+pub struct DriverRegistration {
+    this_module: &'static crate::ThisModule,
+    registered: bool,
+    name: CStr<'static>,
+    probe: Option<SpiMethod>,
+    remove: Option<SpiMethod>,
+    shutdown: Option<SpiMethodVoid>,
+    spi_driver: Option<bindings::spi_driver>,
+}
+
+impl DriverRegistration {
+    fn new(
+        this_module: &'static crate::ThisModule,
+        name: CStr<'static>,
+        probe: Option<SpiMethod>,
+        remove: Option<SpiMethod>,
+        shutdown: Option<SpiMethodVoid>,
+    ) -> Self {
+        DriverRegistration {
+            this_module,
+            name,
+            registered: false,
+            probe,
+            remove,
+            shutdown,
+            spi_driver: None,
+        }
+    }
+
+    // FIXME: Add documentation
+    pub fn new_pinned(
+        this_module: &'static crate::ThisModule,
+        name: CStr<'static>,
+        probe: Option<SpiMethod>,
+        remove: Option<SpiMethod>,
+        shutdown: Option<SpiMethodVoid>,
+    ) -> Result<Pin<Box<Self>>> {
+        let mut registration = Pin::from(Box::try_new(Self::new(
+            this_module,
+            name,
+            probe,
+            remove,
+            shutdown,
+        ))?);
+
+        registration.as_mut().register()?;
+
+        Ok(registration)
+    }
+
+    // FIXME: Add documentation
+    pub fn register(self: Pin<&mut Self>) -> Result {
+        let mut spi_driver = bindings::spi_driver::default();
+        spi_driver.driver.name = self.name.as_ptr() as *const c_types::c_char;
+        spi_driver.probe = self.probe;
+        spi_driver.remove = self.remove;
+        spi_driver.shutdown = self.shutdown;
+
+        let this = unsafe { self.get_unchecked_mut() };
+        if this.registered {
+            return Err(Error::EINVAL);
+        }
+
+        this.spi_driver = Some(spi_driver);
+
+        let res = unsafe {
+            bindings::__spi_register_driver(this.this_module.0, this.spi_driver.as_mut().unwrap())
+        };
+
+        match res {
+            0 => {
+                this.registered = true;
+                Ok(())
+            }
+            _ => Err(Error::from_kernel_errno(res)),
+        }
+    }
+}
+
+impl Drop for DriverRegistration {
+    fn drop(&mut self) {
+        unsafe { bindings::driver_unregister(&mut self.spi_driver.as_mut().unwrap().driver) }
+        // FIXME: No unwrap? But it's safe?
```

Comment:
Does `bindings::spi_driver` need to live inside an `Option`? If you could get rid of the `Option`, you would not need `unwrap()` here (which could theoretically panic).

Reply:
> Does bindings::spi_driver need to live inside an Option

I think not. I made a [PR](https://github.com/ks0n/linux/pull/9) but I need to see if it works correctly before merging it